### ISSUE TITLE
DPoP makes unrealistic claims about client time

### ISSUE THREAD

**Issue by will-bartlett** [danielfett/draft-dpop#67]
> If an adversary is able to get hold of a DPoP proof JWT, the adversary could replay that token at the same endpoint (the HTTP endpoint and method are enforced via the respective claims in the JWTs). To prevent this, servers MUST only accept DPoP proofs for a limited time window after their iat time, preferably only for a relatively brief period (on the order of a few seconds)

Time skew between client devices like mobile phones and cloud services is routinely greater than a few seconds. This is easy to see - choose your favorite cloud provider, spin up a VM in each region, and ask them all the time - they won't be within a few seconds.  A few minutes is a more realistic statement, unless the client is using the server timestamp to compute and account for skew, and then only if the client and server have affinity after time sync.

**Comment by selfissued** [comment]
Like #69, I believe this is largely fixed by Largely fixed by PR #81.

**Comment by selfissued** [comment]
We could add text to the spec saying that if a server nonce is used, liveness checks could be performed at the server using nonce state rather than only "iat".

**Comment by will-bartlett** [comment]
I'm not so sure this is fixed by #81. For Microsoft, #81 fixes the issue - we'll use nonce, we'll ignore "iat", all good. But, the spec still makes an unrealistic claim by saying "a few seconds" - other implementors, who want to use client time, will try this and find it doesn't work reliably.

I would still recommend removing the parenthetical or changing the parenthetical from "(on the order of a few seconds)" to "(on the order of a minute)"

**Comment by PieterKas** [comment]
> We could add text to the spec saying that if a server nonce is used, liveness checks could be performed at the server using nonce state rather than only "iat".

Section 10.1 already has language to that effect, but it is a bit hidden, perhaps it can broken into its own paragraph?

   Note: To accommodate for clock offsets, the server MAY accept DPoP
   proofs that carry an iat time in the reasonably near future (e.g., a
   few seconds in the future).  **Because clock skews between servers and
   clients may be large, servers may choose to limit DPoP proof
   lifetimes by using server-provided nonce values rather than clock
   times, yielding intended results even in the face of arbitrarily
   large clock skews.**

**Comment by PieterKas** [comment]
> 



> I'm not so sure this is fixed by #81. For Microsoft, #81 fixes the issue - we'll use nonce, we'll ignore "iat", all good. But, the spec still makes an unrealistic claim by saying "a few seconds" - other implementors, who want to use client time, will try this and find it doesn't work reliably.
> 
> I would still recommend removing the parenthetical or changing the parenthetical from "(on the order of a few seconds)" to "(on the order of a minute)"

Specifying something to be "on the order" feels imprecise. Omitting it completely creates no expectation either. Could we specify an upper bound? e.g.

To prevent this, servers MUST only accept DPoP proofs for a limited time window after their iat time, preferably only for a relatively brief period (**no more than one minute**). **If a one minute time window cannot be achieved, using a server supplied nonce is recommended**

**Comment by selfissued** [comment]
The draft already includes clock handling text equivalent to what the editors discussed having me add.  Specifically, we already say:
> Because clock skews between servers and clients may be large, servers may choose to limit DPoP proof lifetimes by using server-provided nonce values rather than clock times, yielding intended results even in the face of arbitrarily large clock skews.

I propose closing this issue on this basis, because I believe its intent is already addressed in the specification.

**Comment by bc-pi** [comment]
I tend to agree with @will-bartlett that the one parenthetical should be removed or adjusted.
